chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,195 @@
+        alpha (float): Weight for variance loss    (alpha * variance_loss)
+        beta (float): Weight for distance loss     (beta * distance_loss)
+        gamma (float): Weight for regularizer loss (gamma * regularizer_loss)
+    """
+
+    def __init__(self, delta_v=0.5, delta_d=1.5,
+                 max_embedding_dim=10, norm=1,
+                 alpha=1.0, beta=1.0, gamma=0.001):
+        self.delta_v = delta_v
+        self.delta_d = delta_d
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+        self.max_embedding_dim = max_embedding_dim
+
+        if self.max_embedding_dim <= 0:
+            raise ValueError("Max number of embeddings has to be positive!")
+
+        # L1 or L2 norm is allowed only
+        if norm == 1:
+            self.norm = lambda x, axis=None: c_sum(absolute(x), axis=axis)
+        elif norm == 2:
+            self.norm = lambda x, axis=None: sqrt(c_sum(x ** 2, axis=axis))
+        else:
+            raise ValueError("For discriminative loss, "
+                             "norm can only be 1 or 2. "
+                             "Obtained the value : {}".format(norm))
+
+    def __call__(self, embeddings, labels):
+
+        """Applies discriminative margin based clustering loss
+
+        Args:
+            embeddings(:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`) :
+                predicted embedding vectors
+                (batch size, max embedding dimension, height, width)
+            labels(:class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`) :
+                instance segmentation ground truth
+                each unique value has to be denoting one instance
+                (batch size, height, width)
+
+        Returns:
+            Variance Loss(:class:`~chainer.Variable` or float):
+                variance loss multiplied by alpha
+            Distance Loss(:class:`~chainer.Variable` or float):
+                distance loss multiplied by beta
+            Regularization Loss(:class:`~chainer.Variable` or float):
+                regularization loss multiplied by gamma
+        """
+
+        assert (self.max_embedding_dim == embeddings.shape[1])
+
+        l_dist = 0.0
+        count = 0
+        xp = cuda.get_array_module(embeddings)
+
+        emb = embeddings[None, :]
+        emb = broadcast_to(emb, (emb.shape[1],
+                                 emb.shape[1],
+                                 emb.shape[2],
+                                 emb.shape[3],
+                                 emb.shape[4]))
+        ms = []
+        for c in range(self.max_embedding_dim):
+            # Create mask for instance
+            mask = xp.expand_dims(labels == c + 1, 1)
+            ms.append(mask)
+        ms = xp.stack(ms, 0)
+        mns = c_sum(emb * ms, axis=(3, 4))
+        mns = mns / xp.maximum(xp.sum(ms, (2, 3, 4))[:, :, None], 1)
+        mns_exp = mns[:, :, :, None, None]
```

Comment:
Just for confirmation. Cluster means are functions of `embeddings`, but we don't have to propagate error via means, do we?